src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/ScaleToTarget.java

```diff
@@ -0,0 +1,416 @@
+
+/**
+ * Class encapsulating the logic to scale a coverage to a pre-defined target size.
+ *
+ * @author Stefano Costa, GeoSolutions
+ *
+ */
+class ScaleToTarget {
+
+    /** The overview policy. By default, NEAREST policy is used **/
+    private OverviewPolicy overviewPolicy;
+
+    /** The interpolation method. By default, NEAREST interpolation is used **/
+    private Interpolation interpolation;
+
+    private GridCoverage2DReader reader;
+
+    private Envelope envelope;
+
+    private Integer adjustedTargetSizeX;
+
+    private Integer adjustedTargetSizeY;
+
+    /**
+     * Constructor.
+     *
+     * @param reader the coverage reader to use for reading metadata
+     */
+    ScaleToTarget(GridCoverage2DReader reader) {
+        this(reader, null);
+    }
+
+    /**
+     * Two-args constructor.
+     *
+     * @param reader the coverage reader to use for reading metadata
+     * @param envelope the envelope of the ROI we want to scale (if <code>null</code>, the envelope of the whole coverage is used)
+     */
+    ScaleToTarget(GridCoverage2DReader reader, Envelope envelope) {
+        checkNotNull(reader, "reader");
+        this.reader = reader;
+        this.envelope = envelope;
+        if (this.envelope == null) {
+            this.envelope = reader.getOriginalEnvelope();
+        }
+        this.interpolation = DownloadUtilities.INTERPOLATION_NEAREST;
+        this.overviewPolicy = OverviewPolicy.NEAREST;
+    }
+
+    /**
+     * @return the interpolation
+     */
+    public Interpolation getInterpolation() {
+        return interpolation;
+    }
+
+    /**
+     * @param interpolation the interpolation to set
+     */
+    public void setInterpolation(Interpolation interpolation) {
+        checkNotNull(interpolation, "interpolation");
+        this.interpolation = interpolation;
+    }
+
+    /**
+     * @return the overviewPolicy
+     */
+    public OverviewPolicy getOverviewPolicy() {
+        return overviewPolicy;
+    }
+
+    /**
+     * @param overviewPolicy the overviewPolicy to set
+     */
+    public void setOverviewPolicy(OverviewPolicy overviewPolicy) {
+        checkNotNull(overviewPolicy, "overviewPolicy");
+        this.overviewPolicy = overviewPolicy;
+    }
+
+    /**
+     * @return the current target size
+     */
+    public Integer[] getTargetSize() {
+        return new Integer[] { this.adjustedTargetSizeX, this.adjustedTargetSizeY };
+    }
+
+    /**
+     * Sets the size of the scaled image (target).
+     *
+     * <p>
+     * If one of the two inputs is omitted, the missing value is inferred from the provided one so that the aspect ratio of the specified envelope is
+     * preserved.
+     * </p>
+     *
+     * @param targetSizeX the size of the target image along the X axis
+     * @param targetSizeY the size of the target image along the Y axis
+     * @throws TransformException
+     */
+    public void setTargetSize(Integer targetSizeX, Integer targetSizeY) throws TransformException {
```

Comment:
Agreed, but this method will not be called often and using `Integer` allows to deal with missing values more easily.